File: src/model/Deck.java
package model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

/**
 * Represents a deck of cards that can be shuffled and drawn from.
 */
public class Deck {
    private Stack<Card> cards;

    public Deck() {
        cards = new Stack<>();
    }

    /**
     * Creates a standard 52-card deck.
     */
    public static Deck createStandardDeck() {
        Deck deck = new Deck();
        for (Suit suit : Suit.values()) {
            for (Rank rank : Rank.values()) {
                deck.cards.push(new Card(suit, rank));
            }
        }
        return deck;
    }

    /**
     * Creates multiple standard decks combined.
     * @param numDecks Number of 52-card decks to combine
     */
    public static Deck createMultipleDecks(int numDecks) {
        Deck deck = new Deck();
        for (int i = 0; i < numDecks; i++) {
            for (Suit suit : Suit.values()) {
                for (Rank rank : Rank.values()) {
                    deck.cards.push(new Card(suit, rank));
                }
            }
        }
        return deck;
    }

    /**
     * Shuffles the deck randomly.
     */
    public void shuffle() {
        List<Card> cardList = new ArrayList<>(cards);
        Collections.shuffle(cardList);
        cards.clear();
        for (Card card : cardList) {
            cards.push(card);
        }
    }

    /**
     * Draws a card from the top of the deck.
     * @return The top card, or null if deck is empty
     */
    public Card draw() {
        if (cards.isEmpty()) {
            return null;
        }
        return cards.pop();
    }

    /**
     * Draws multiple cards from the deck.
     * @param count Number of cards to draw
     * @return List of drawn cards (may be less than count if deck runs out)
     */
    public List<Card> draw(int count) {
        List<Card> drawn = new ArrayList<>();
        for (int i = 0; i < count && !cards.isEmpty(); i++) {
            drawn.add(cards.pop());
        }
        return drawn;
    }

    /**
     * Returns the top card without removing it.
     * @return The top card, or null if deck is empty
     */
    public Card peek() {
        if (cards.isEmpty()) {
            return null;
        }
        return cards.peek();
    }

    /**
     * Adds a card to the top of the deck.
     */
    public void addCard(Card card) {
        cards.push(card);
    }

    /**
     * Adds multiple cards to the deck.
     */
    public void addCards(List<Card> cardsToAdd) {
        for (Card card : cardsToAdd) {
            cards.push(card);
        }
    }

    /**
     * Returns the number of cards remaining in the deck.
     */
    public int size() {
        return cards.size();
    }

    /**
     * Returns true if the deck is empty.
     */
    public boolean isEmpty() {
        return cards.isEmpty();
    }

    /**
     * Clears all cards from the deck.
     */
    public void clear() {
        cards.clear();
    }
}

File: 21CardGame/mobile/src/models/Deck.ts
import { Card, createCard } from './Card';
import { Suit, Rank, getAllSuits, getAllRanks } from './types';

/**
 * Represents a deck of cards that can be shuffled and drawn from.
 * Uses an array as a stack (end of array = top of deck)
 */
export interface Deck {
  cards: Card[];
}

/**
 * Create an empty deck
 */
export function createEmptyDeck(): Deck {
  return { cards: [] };
}

/**
 * Creates a standard 52-card deck
 */
export function createStandardDeck(): Deck {
  const cards: Card[] = [];
  for (const suit of getAllSuits()) {
    for (const rank of getAllRanks()) {
      cards.push(createCard(suit, rank));
    }
  }
  return { cards };
}

/**
 * Creates multiple standard decks combined.
 * @param numDecks Number of 52-card decks to combine
 */
export function createMultipleDecks(numDecks: number): Deck {
  const cards: Card[] = [];
  for (let i = 0; i < numDecks; i++) {
    for (const suit of getAllSuits()) {
      for (const rank of getAllRanks()) {
        cards.push(createCard(suit, rank));
      }
    }
  }
  return { cards };
}

/**
 * Fisher-Yates shuffle algorithm
 */
export function shuffleDeck(deck: Deck): Deck {
  const cards = [...deck.cards];
  for (let i = cards.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [cards[i], cards[j]] = [cards[j], cards[i]];
  }
  return { cards };
}

/**
 * Draw a card from the top of the deck
 * @returns [drawnCard, remainingDeck] or [null, deck] if empty
 */
export function drawCard(deck: Deck): [Card | null, Deck] {
  if (deck.cards.length === 0) {
    return [null, deck];
  }
  const cards = [...deck.cards];
  const card = cards.pop()!;
  return [card, { cards }];
}

/**
 * Draw multiple cards from the deck
 * @param count Number of cards to draw
 * @returns [drawnCards, remainingDeck]
 */
export function drawCards(deck: Deck, count: number): [Card[], Deck] {
  const drawn: Card[] = [];
  const cards = [...deck.cards];
  for (let i = 0; i < count && cards.length > 0; i++) {
    drawn.push(cards.pop()!);
  }
  return [drawn, { cards }];
}

/**
 * Peek at the top card without removing it
 */
export function peekDeck(deck: Deck): Card | null {
  if (deck.cards.length === 0) {
    return null;
  }
  return deck.cards[deck.cards.length - 1];
}

/**
 * Add a card to the top of the deck
 */
export function addCardToDeck(deck: Deck, card: Card): Deck {
  return { cards: [...deck.cards, card] };
}

/**
 * Add multiple cards to the deck
 */
export function addCardsToDeck(deck: Deck, cardsToAdd: Card[]): Deck {
  return { cards: [...deck.cards, ...cardsToAdd] };
}

/**
 * Get the number of cards in the deck
 */
export function deckSize(deck: Deck): number {
  return deck.cards.length;
}

/**
 * Check if the deck is empty
 */
export function isDeckEmpty(deck: Deck): boolean {
  return deck.cards.length === 0;
}

/**
 * Clear all cards from the deck
 */
export function clearDeck(): Deck {
  return { cards: [] };
}
